x/cosmwasmpool/README.md

```diff
@@ -0,0 +1,284 @@
+
+```mermaid
+graph TD;
+  Sender((Sender))
+
+  Sender -- create poool --> x/cosmwasmpool
+  x/cosmwasmpool -- get next & set pool id --> x/poolmanager
+  x/cosmwasmpool -- instantiate contract --> x/wasm
+```
+
+The CosmWasm contract that is to be instanitiated needs to implement [CosmWasm Pool Contract Interface](#cosmwasm-pool-contract-interface) and store it on chain first. Then new pool can be created by sending `MsgCreateCosmWasmPool`.
+
+
+`MsgCreateCosmWasmPool` contains `InstantiateMsg`, which is a message that will be passed to the CosmWasm contract when it is instantiated. The structure of the message is defined by the contract developer, and can contain any information that the contract needs to be instantiated. JSON format is used for `InstantiateMsg`.
+
+```mermaid
+sequenceDiagram
+    participant Sender
+    participant x/cosmwasmpool
+    participant x/wasm
+    participant x/poolmanager
+
+    Sender ->> x/cosmwasmpool: MsgCreateCosmWasmPool {CodeId, InstantiateMsg, Sender}
+
+    Note over x/wasm: Given there is a pool contract with CodeId
+
+
+    x/cosmwasmpool ->> x/wasm: Call InstantiateContract(CodeId, InstantiateMsg)
+    x/wasm -->> x/cosmwasmpool: ContractAddress
+    x/poolmanager ->> x/cosmwasmpool: Call GetNextPoolId()
+    x/cosmwasmpool ->> x/poolmanager: Call SetNextPoolId(poolId)
+
+    Note over x/cosmwasmpool: Store CodeId, ContractAddress, and PoolId
+
+    x/cosmwasmpool -->>  Sender: MsgCreateCosmWasmPoolResponse {PoolId}
+```
+
+
+## Providing / Withdrawing Liquidity
+
+Currently, all existing pool types has their own way of providing liquidity and shares calculation. CosmWasm pool aims to be flexible that regards and let the contract define the way of providing liquidity. So there is no restriction here, and the contract developer can define the way of providing liquidity as they wish, potentially with execute endpoint since `MsgExecuteContract` triggers state mutating endpoint and can also attach funds to it.
+
+Common interface and later be defined for the contract to implement as spec and/or create a separated crate for that purpose.
+
+It's important to note that the _**contract itselfs hold tokens that are provided by users**_.
+
+
+## Swap
+
+One of the main reason why CosmWasm pool is implemented as a module + contract rather than a contract only is that it allows us to use the existing pool manager module to handle swap, which means things like swap routing, cross chain swap, and other functionality that depends on existing pool interface works out of the box.
+
+```mermaid
+graph TD;
+  Sender((Sender))
+  Sender -- swap --> x/poolmanager
+  x/poolmanager -- route msg to --> x/cosmwasmpool
+  x/cosmwasmpool -- sudo execute contract --> x/wasm
+  x/wasm -- sudo --> wasm/pool
+
+  x/cosmwasmpool -- send token_in from sender to wasm/pool --> x/bank
+  wasm/pool -- send token_out to sender --> x/bank
```

Comment:
From the diagram it's unclear to me where ths bank send happens. Could you please also elaborate on the latest decisions with respect to bank send? Do they happen on chain and why or why not?

Reply:
I will add sequence diagram for clarifying this part then

Reply:
Confirmed that this was added